### PR TITLE
shadow: add tcb support

### PR DESCRIPTION
**Pull request by Izorkin**
###### Description of changes
Add support alternative password shadowing scheme.
Issue - https://github.com/NixOS/nixpkgs/issues/109457

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Izorkin** [comment]
cc @SuperSandro2000 @matthewbauer @NeQuissimus @alyssais

File: pkgs/os-specific/linux/shadow/default.nix [hunk 1]
```diff
@@ -1,6 +1,6 @@
 { lib, stdenv, nixosTests, fetchpatch, fetchFromGitHub, autoreconfHook, libxslt
 , libxml2 , docbook_xml_dtd_45, docbook_xsl, itstool, flex, bison, runtimeShell
-, libxcrypt, pam ? null, glibcCross ? null
+, libxcrypt, pam ? null, tcb ? null, glibcCross ? null
```
[the remark this answers is not in the document]
Comment (Izorkin) [reply pkgs/os-specific/linux/shadow/default.nix:3]: I thought I had to do the same thing here as with the `libxcrypt` and `glibcCross` packages.

**Comment by Izorkin** [comment]
Rebased PR.

File: pkgs/os-specific/linux/shadow/default.nix [hunk 2]
```diff
@@ -1,6 +1,7 @@
 { lib, stdenv, nixosTests, fetchpatch, fetchFromGitHub, autoreconfHook, libxslt
 , libxml2 , docbook_xml_dtd_45, docbook_xsl, itstool, flex, bison, runtimeShell
 , libxcrypt, pam ? null, glibcCross ? null
+, withTcb ? true, tcb
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/shadow/default.nix:4]:
```suggestion
, withTcb ? stdenv.isLinux, tcb
```

File: pkgs/os-specific/linux/shadow/default.nix [hunk 3]
```diff
@@ -24,7 +25,8 @@ stdenv.mkDerivation rec {
   };
 
   buildInputs = [ libxcrypt ]
-    ++ lib.optional (pam != null && stdenv.isLinux) pam;
+    ++ lib.optional (pam != null && stdenv.isLinux) pam
+    ++ lib.optional (withTcb && stdenv.isLinux) tcb;
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/shadow/default.nix:29]:
```suggestion
    ++ lib.optional withTcb tcb;
```

File: pkgs/os-specific/linux/shadow/default.nix [hunk 4]
```diff
@@ -24,7 +25,8 @@ stdenv.mkDerivation rec {
...
   nativeBuildInputs = [autoreconfHook libxslt libxml2
     docbook_xml_dtd_45 docbook_xsl flex bison itstool
     ];
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/shadow/default.nix:32]:
```suggestion
  nativeBuildInputs = [
    autoreconfHook libxslt libxml2
    docbook_xml_dtd_45 docbook_xsl flex bison itstool
  ];
```

unrelated

**Comment by Izorkin** [comment]
@SuperSandro2000 thanks, fixed.
Did a little cleanup.
